src/xdp/program.c

```diff
@@ -1312,6 +1312,38 @@ XdpProgramTraceObject(
     }
 }
 
+static
+_IRQL_requires_max_(DISPATCH_LEVEL)
+VOID
+XdpProgramCompileUpdateProgram(
+    _In_ XDP_RX_QUEUE *RxQueue
+    )
+{
+    LIST_ENTRY *BindingListHead = XdpRxQueueGetProgramBindingList(RxQueue);
+    XDP_PROGRAM *Program = XdpRxQueueGetProgram(RxQueue);
+    LIST_ENTRY *Entry = BindingListHead->Flink;
+
+    TraceEnter(TRACE_CORE, "Updating program on RxQueue=%p", RxQueue);
+
+    Program->RuleCount = 0;
+
+    while (Entry != BindingListHead) {
+        XDP_PROGRAM_BINDING* ProgramBinding =
+            CONTAINING_RECORD(Entry, XDP_PROGRAM_BINDING, RxQueueEntry);
+        CONST XDP_PROGRAM_OBJECT *BoundProgramObject = ProgramBinding->OwningProgram;
+
+        for (UINT32 i = 0; i < BoundProgramObject->Program.RuleCount; i++) {
+            Program->Rules[Program->RuleCount++] = BoundProgramObject->Program.Rules[i];
```

Comment:
Can we add an assert here for Program->RuleCount being in bounds?

Reply:
Added a check and hit it immediately :/

Reply:
Ok, fixed the bug found by the assert.